tooling/test/sugar.go

```diff
@@ -2,58 +2,143 @@ package test
+func TestURLFromSpecURL(testURL string) string {
+	u, err := url.Parse(testURL)
+	if err != nil {
+		panic(err)
+	}
+
+	withoutExampleDomain := strings.TrimSuffix(u.Host, GATEWAY_LOCALHOST_DOMAIN)
+
+	if len(withoutExampleDomain) < len(u.Host) {
+		// we found and removed the example domain
+		// so this is a subdomain gateway
+		u.Scheme = SubdomainGatewayScheme
+		u.Host = fmt.Sprintf("%s%s", withoutExampleDomain, SubdomainGatewayHost)
+	}
+
+	return u.String()
+}
+
+func ReplaceExampleDomain(s string, args ...any) string {
+	return TestURLFromSpecURL(fmt.Sprintf(s, args...))
+}
+
 type RequestBuilder struct {
-	Method_  string
-	Url_     string
-	Headers_ map[string]string
+	Method_               string
+	Path_                 string
+	URL_                  string
+	Proxy_                string
+	UseProxyTunnel        bool
+	Headers_              map[string]string
+	DoNotFollowRedirects_ bool
+	Query_                url.Values
 }
 
 func Request() RequestBuilder {
-	return RequestBuilder{Method_: "GET"}
+	return RequestBuilder{Method_: "GET",
+		Query_: make(url.Values)}
+}
+
+func (r RequestBuilder) Path(path string, args ...any) RequestBuilder {
+	r.Path_ = fmt.Sprintf(path, args...)
+	return r
 }
 
-func (r RequestBuilder) Url(url string, args ...any) RequestBuilder {
-	r.Url_ = fmt.Sprintf(url, args...)
+func (r RequestBuilder) URL(path string, args ...any) RequestBuilder {
+	r.URL_ = fmt.Sprintf(path, args...)
 	return r
 }
 
-func (r RequestBuilder) Header(h HeaderBuilder) RequestBuilder {
+func (r RequestBuilder) Query(key, value string, args ...any) RequestBuilder {
+	r.Query_.Add(key, fmt.Sprintf(value, args...))
+	return r
+}
+
+func (r RequestBuilder) GetURL() string {
+	if r.Path_ != "" {
+		panic("not supported")
+	}
+
+	return r.URL_
+}
+
+func (r RequestBuilder) Proxy(path string, args ...any) RequestBuilder {
+	r.Proxy_ = fmt.Sprintf(path, args...)
+	return r
+}
+
+func (r RequestBuilder) WithProxyTunnel() RequestBuilder {
+	r.UseProxyTunnel = true
+	return r
+}
+
+func (r RequestBuilder) DoNotFollowRedirects() RequestBuilder {
+	r.DoNotFollowRedirects_ = true
+	return r
+}
+
+func (r RequestBuilder) Method(method string) RequestBuilder {
+	r.Method_ = method
+	return r
+}
+
+func (r RequestBuilder) Header(k, v string) RequestBuilder {
 	if r.Headers_ == nil {
 		r.Headers_ = make(map[string]string)
 	}
 
-	r.Headers_[h.Key] = h.Value
+	r.Headers_[k] = v
 	return r
 }
 
 func (r RequestBuilder) Headers(hs ...HeaderBuilder) RequestBuilder {
+	if r.Headers_ == nil {
+		r.Headers_ = make(map[string]string)
+	}
+
 	for _, h := range hs {
-		r = r.Header(h)
+		r.Headers_[h.Key] = h.Value
 	}
 
 	return r
 }
 
 func (r RequestBuilder) Request() CRequest {
+	if r.URL_ != "" && r.Path_ != "" {
+		panic("Both 'Url' and 'Path' are set")
+	}
+
 	return CRequest{
-		Method:  r.Method_,
-		Url:     r.Url_,
-		Headers: r.Headers_,
+		Method:               r.Method_,
+		Path:                 r.Path_,
+		Url:                  r.URL_,
```

Comment:
Could we make this `URL` too for consistency?

Reply:
I'll rename all `Url` names into `URL`, thanks for the catch.